<!DOCTYPE html PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN">
<html>
  <head>
    <meta http-equiv="content-type" content="text/html; charset=UTF-8">
    <title>PiiL tutorial</title>
  </head>
  <body>
    <br>
    <div align="center"><img alt="logo"
        src="../Documents/Manfred/graphics/pill-manual.png" height="173"
        width="628"><br>
    </div>
    <h1 align="left">Welcome to PiiL!</h1>
    <h1> </h1>
    PiiL is an open-source (<b>the source code is available </b><b><a
        href="https://github.com/behroozt/PiiL.git">here</a></b>) and
    biologist-friendly tool. You can easily <i>PiiL</i> your data; i.e.
    load your DNA methylation and/or RNA expression data over any of the
    KEGG pathways and let PiiL color-code it for you to recognize or
    discover the patterns of differential methylation / expression for
    your targeted genes. PiiL provides statistical graphs for the
    matched genes in the pathway. With PiiL you can also produce vector
    quality images of the pathways and their visualized loaded data or a
    list of genes highlighted in the pathway that are suitable to be
    used in publications.<br>
    <p> <i>PiiL </i>your data and cite us, please.<br>
    </p>
    <h1><a name="How_to"></a>How to?</h1>
    <blockquote>
      <h3><a name="Open_a_KGML_file"></a></h3>
      <h3><a href="#Run_PiiL_">Run PiiL</a><br>
      </h3>
      <h3><a href="#Open_a_KGML_file_">Open a KGML file</a></h3>
      <h3><a href="#Load_methylation_data_">Load methylation data</a></h3>
      <h3><a href="#Load_expression_data_">Load expression data</a></h3>
      <h3><a href="#Load_samples_information_">Load samples information</a></h3>
      <h3><a href="#Right-click_menu_options_">Right-click menu options</a><br>
      </h3>
      <h3><a href="#Highlight_a_set_of_genes_in_a_pathway_">Highlight a
          set of genes in a pathway</a></h3>
      <h3><a href="#Export_pathway_to_image_">Export pathway to image</a></h3>
      <h3><a href="#Export_list_of_matched_genes_in_each_">Export list
          of matched genes in each pathway</a></h3>
      <h3><a href="#Duplicate_the_current_pathway_in_a_new_">Duplicate
          the current pathway in a new tab</a></h3>
      <h3><a href="#Taking_a_snapshot">Take a snapshot from PiiL</a><br>
      </h3>
      <h3><a href="#Cite_us_">Cite us</a></h3>
      <h3><a href="#What_the_name_PiiL_means">Know what the name PiiL
          means</a><br>
      </h3>
    </blockquote>
    <p><br>
    </p>
    <h2><a name="Run_PiiL_"></a>Run PiiL</h2>
    <p>To run PiiL on Mac just double-click on the PiiL.jar file. For
      Windows and Linux platforms, open a terminal, go to the directory
      that you have PiiL.jar file and run "java -jar PiiL.jar".<br>
    </p>
    <h3><a href="#How_to">back to top</a></h3>
    <p><br>
    </p>
    <h2> </h2>
    <h2><a name="Open_a_KGML_file_"></a>Open a KGML file</h2>
    <p>There are two options to open a KGML file:<br>
      <img alt="open KGML" src="images/openKGML.png" height="700"
        width="1050"><br>
    </p>
    <p>1) Open an already downloaded KGML file from your hard drive.<br>
      2) Open it from by selecting the pathway and organism name from
      the KEGG database. This will download the file and for the next
      time will use the downloaded file.<br>
      <img alt="open from web" src="images/openFromWeb.png" height="250"
        width="910"><br>
    </p>
    <h3><a href="#How_to">back to top</a></h3>
    <h6> </h6>
    <h2><a name="Load_methylation_data_"></a>Load methylation data</h2>
    <p>Having open a pathway, you can load your meta-data over that
      pathway. If you have a chip's methylation data, replace the CpG
      sites in the first column with their annotated gene name.
      Additional annotation data (for example if the CpG site lies in a
      exonic, intronic, ... region of the gene) can come after an
      underscore in the first column. Columns must be tab separated and
      the first row lists the samples IDs. The beta values are between 0
      and 1.<br>
      A valid input file lookes like the following:<br>
      <br>
    </p>
    <table height="142" border="0" cellpadding="2" cellspacing="2"
      width="676">
      <tbody>
        <tr>
          <td valign="top">sampleIDS<br>
          </td>
          <td valign="top">sample1&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;





























            <br>
          </td>
          <td valign="top">sample2&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;





























            <br>
          </td>
          <td valign="top">...&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;





























            <br>
          </td>
          <td valign="top">sampleN<br>
          </td>
        </tr>
        <tr>
          <td valign="top">GeneA[_exonic]<br>
          </td>
          <td valign="top">beta_value<br>
          </td>
          <td valign="top">beta_value</td>
          <td valign="top">...<br>
          </td>
          <td valign="top">beta_value</td>
        </tr>
        <tr>
          <td valign="top">GeneB[_intronic]<br>
          </td>
          <td valign="top">beta_value</td>
          <td valign="top">beta_value</td>
          <td valign="top">...<br>
          </td>
          <td valign="top">beta_value</td>
        </tr>
        <tr>
          <td valign="top">...<br>
          </td>
          <td valign="top">...<br>
          </td>
          <td valign="top">...<br>
          </td>
          <td valign="top">...<br>
          </td>
          <td valign="top">...<br>
          </td>
        </tr>
        <tr>
          <td valign="top">GeneX[_upstream]<br>
          </td>
          <td valign="top">beta_value<br>
          </td>
          <td valign="top">beta_value<br>
          </td>
          <td valign="top">...<br>
          </td>
          <td valign="top">beta_value<br>
          </td>
        </tr>
      </tbody>
    </table>
    <p><br>
      To load your methylation data, you can either click on New or any
      of previously loaded methylation files listed under <b><big>Load
          -&gt; Methylation</big>.</b><br>
    </p>
    <img alt="load methylation" src="images/loadMethylation.png"
      height="700" width="1050"><br>
    The genes in the input file that match with the genes in the pathway
    are colored according to their beta values. Blue color codes the
    hypo-methylation status (the closer the beta value is to zero the
    color is darker blue) and color red color codes the
    hyper-methylation status (the closer the beta value is to 1 the
    color is darker red).<br>
    The samples IDs are listed in the first combobox and the matched
    genes are listed in the second combobox where the number of matches
    is shown by the lable below it.<br>
    <br>
    <h3><a href="#How_to">back to top</a></h3>
    <p><br>
    </p>
    <h2><a name="Load_expression_data_"></a>Load expression data</h2>
    <p>Genes and their expression values are the second type of
      meta-data that you can load on your pathway of interest. Make sure
      the input file is tab separated with sample IDs on the first row
      and gene names on the first column. For each gene, the expression
      values appear on its following columns. <br>
      A valid input file lookes like the following:<br>
      <br>
    </p>
    <table height="142" border="0" cellpadding="2" cellspacing="2"
      width="676">
      <tbody>
        <tr>
          <td valign="top">sampleIDS&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
            &nbsp; &nbsp;&nbsp; &nbsp;&nbsp;&nbsp; <br>
          </td>
          <td valign="top">sample1&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;





























            &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;
            &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
            <br>
          </td>
          <td valign="top">sample2&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
            &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;





























            <br>
          </td>
          <td valign="top">...&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
            &nbsp;&nbsp;&nbsp;&nbsp;&nbsp; <br>
          </td>
          <td valign="top">sampleN<br>
          </td>
        </tr>
        <tr>
          <td valign="top">GeneA<br>
          </td>
          <td valign="top">expression_value<br>
          </td>
          <td valign="top">expression_value</td>
          <td valign="top">...<br>
          </td>
          <td valign="top">expression_value</td>
        </tr>
        <tr>
          <td valign="top">GeneB<br>
          </td>
          <td valign="top">expression_value</td>
          <td valign="top">expression_value</td>
          <td valign="top">...<br>
          </td>
          <td valign="top">expression_value</td>
        </tr>
        <tr>
          <td valign="top">...<br>
          </td>
          <td valign="top">...<br>
          </td>
          <td valign="top">...<br>
          </td>
          <td valign="top">...<br>
          </td>
          <td valign="top">...<br>
          </td>
        </tr>
        <tr>
          <td valign="top">GeneX<br>
          </td>
          <td valign="top">expression_value </td>
          <td valign="top">expression_value </td>
          <td valign="top">...<br>
          </td>
          <td valign="top">expression_value </td>
        </tr>
      </tbody>
    </table>
    <p><br>
      To load your gene expression data, you can either click on New or
      any of previously loaded expression files listed under <b><big>Load





























          -&gt; Gene Expression</big>.</b><br>
    </p>
    <br>
    <img alt="load expression" src="images/loadExpression.png"
      height="700" width="1050"> <br>
    <br>
    The genes in the input file that match with the genes in the pathway
    are colored according to their beta values. Blue color codes the
    hypo-methylation status (the closer the beta value is to zero the
    color is darker blue) and color red color codes the
    hyper-methylation status (the closer the beta value is to 1 the
    color is darker red).<br>
    The samples IDs are listed in the first combobox and the matched
    genes are listed in the second combobox where the number of matches
    is shown by the lable below it.<br>
    <h3><br>
    </h3>
    <h3><a href="#How_to">back to top</a></h3>
    <p><br>
    </p>
    <h2><a name="Load_samples_information_"></a>Load samples information</h2>
    <p>If the samples in the loaded meta-data files have additional
      information like gender, age, ehtnicity, etc. you can load the
      file with matched sample IDs and the additional data will be shown
      for each sample.<br>
      Samples information can be loaded from <b><big>Load -&gt; Samples
          information</big> </b>menu, by choosing new or a previously
      loaded samples information file. In the form that shows up choose
      the columns separator, number of fileds (columns), the column
      containing sample IDs, and the fileds that you like to be shown.<br>
    </p>
    <img alt="samples info dialog" src="images/samplesInfo.png"
      height="427" width="567"><br>
    If a valid file is loaded and the samples IDs in the input file
    matches with the ones loaded on the pathway, the additional
    information for each sample is displayed when you navigate through
    samples.<br>
    <br>
    <img alt="sample info example" src="images/samplesInfoExample.png"
      height="700" width="1050"><br>
    <br>
    <h3><a href="#How_to">back to top</a></h3>
    <p><br>
    </p>
    <h2><a name="Right-click_menu_options_"></a>Right-click menu options</h2>
    <p>A menu is available for all the genes when you right-click on the
      gene rectangle. The first option is <b><big>Check this gene on
          ...</big> </b>which is enabled for all the genes. This opens
      a tab in the default browser and checks the target gene's
      information on your selected database: GeneCards, Pubmed or
      Ensemble.</p>
    <img alt="right-click menu options"
      src="../Desktop/PiiL-screenshots/rightClickMenu.png" height="700"
      width="1050">
    <p>The second option <big><b>Histogram for all samples </b></big>makes





























      a histogram for the beta-values or expression values of all
      samples for the targeted gene. <br>
    </p>
    <p><img alt="histogram" src="images/histogram-betaValues.png"
        height="514" width="914">&nbsp;<br>
    </p>
    <p>The third option <b><big>Barplot for raw data</big> </b>is
      another representation of beta-values or expression values for all
      samples.</p>
    <br>
    <br>
    <img alt="barplot" src="images/barplot-betaValues.png" height="514"
      width="914"><br>
    The fourth option Genomic region details is available for the genes
    with a cyan-colored border. A gene with a cyan-colored border show
    that multiple CpG sites have been annotated to this gene and the
    color is the average of all sites. If the genomic region is provided
    in the input data (genomic region can attach the gene name after an
    underscore), you can right click on the gene and all CpG sites for
    this gene - labeled with the genomic region tag - shows up on the
    screen.<br>
    <br>
    <img alt="genomic regions" src="images/genomicRegions.png"
      height="344" width="614"><br>
    <br>
    The last option <big><b>Multiple-sample view</b></big> is explained
    in <a href="#Multiple-sample_view">this</a> section.<br>
    <br>
    <h3><a href="#How_to">back to top</a></h3>
    <p><br>
    </p>
    <h2>Load a list of genes to be highlighted in a target pathway</h2>
    <p>If you have a list of genes and want to check the overlap of this
      list with a pathway of interest, first open the pathway and then
      select <b><big>Load -&gt; List of genes</big> </b>to load your
      file. The genes that exist in the targeted pathway will be
      highlighted with a red text and red border.<br>
    </p>
    <p><img alt="highlight genes" src="images/highlightGenes.png"
        height="700" width="1050"></p>
    <h3><a href="#How_to">back to top</a></h3>
    <br>
    <h2><a name="Export_pathway_to_image_"></a>Export pathway to image</h2>
    The visualized pathway and its color-coded meta data can be exported
    to a vector based image (and other formats). This function is
    available under the<b> <big>Export -&gt; Current pathway to image</big></b>
    menu with two options:<br>
    1) The visible part of the pathway <br>
    2) The entire pathway<br>
    <h3><a href="#How_to">back to top</a></h3>
    <p><br>
    </p>
    <h2><a name="Export_list_of_matched_genes_in_each_"></a>Export list
      of matched genes in each pathway</h2>
    After openning multipe pathways in multiple tabs and loading meta
    data for each of them, this function can be used to get a summary of
    each tab containing the pathway, the meta data loaded over that and
    the list of genes that were matched in each pathway. By clicking on
    <b><big>Export -&gt; List of matched genes in each pathway</big> </b>you





























    can choose a file that keeps this summary.<br>
    &nbsp;<br>
    <h3><a href="#How_to">back to top</a></h3>
    <br>
    <h2><a name="Duplicate_the_current_pathway_in_a_new_"></a>Duplicate
      the current pathway in a new tab</h2>
    Having an open tab with a visualized pathway and meta data loaded,
    you can choose the <b><big>Tools -&gt; Duplicate the current
        pathway</big> </b>option to quickly duplicate the current tab's
    pathway (without the metadata) in a new tab.<br>
    <h3><a href="#How_to">back to top</a></h3>
    <p><br>
    </p>
    <h2><a name="Taking_a_snapshot"></a>Taking a snapshot</h2>
    <p>In addition to the function of exporting pathways and their
      visualized metadata, you can take a snapshot of PiiL in different
      image formats by selecting <b><big>Tools -&gt; Take a snapshot of
          PiiL</big>.<br>
      </b></p>
    <h3><b><a href="#How_to">back to top</a></b></h3>
    <p><br>
    </p>
    <h2><a name="Multiple-sample_view"></a><b>Multiple-sample view</b></h2>
    <p>This function visualizes the metadata for the current sample and
      the next 10 samples, for the selected genes. If the number of
      samples from the current one to the end of the sample set is less
      than 10, as many as possible are displayed.<br>
      To select a gene that is color-coded (i.e. has some metadata
      loaded), double click on the gene rectangle. The selected genes'
      borders turn to yellow. <br>
      Hint: To choose the direction of expansion for multiple-sample
      view, double-click on your desired quarter of the gene rectangle
      while selecting it (image a gene rectangle is divided to four
      equal rectangles). For example double clicking on the north-west
      quarter of the gene sets it to show next samples on the right top
      side of it.<br>
    </p>
    <img alt="multiple-sample view" src="images/multipleSampleView.png"
      height="700" width="1050"><br>
    If you hover the mouse over the extended rectangles, a tooltip shows
    the sample that rectangle is representing.<br>
    You can right-click on any of the expanded genes and choose
    Single-sample view to undo the expansion. To go back to
    Single-sample view for all the expanded genes (the ones in
    multiple-sample view), select <b><big>Tools -&gt; Show
        single-sample view for all genes</big>.</b><br>
    <br>
    <h3><a href="#How_to">back to top</a></h3>
    <p><br>
    </p>
    <h2><a name="Cite_us_"></a>Cite us<br>
    </h2>
    If you <i>PiiL</i> your data please cite our publication ...<br>
    <h3><a href="#How_to">back to top</a></h3>
    <p><br>
    </p>
    <h2><a name="What_the_name_PiiL_means"></a>What the name PiiL means</h2>
    <p>Look at the logo carefully. Do you recognize a cute elephant
      there?! Piil (pronounced as meal) means elephant in Persian (with
      Sanskrit origin). And on a chess board the chessman called bishop
      is actually a <i>piil </i>(elephant) standing next to a hosrse
      (knight)! <br>
    </p>
  </body>
</html>
